netket/graph/_lattice_edge_logic.py

```diff
@@ -0,0 +1,188 @@
+from netket.utils.float import comparable, is_approx_int
+
+from typing import Tuple
+
+# Helper functions
+
+
+def create_site_positions(basis_vectors, extent, site_offsets, extra_shells=None):
+    """Generates the coordinates of all lattice sites.
+
+    extra_shells, if used, must be a vector of the same length as extent"""
+    if extra_shells is None:
+        extra_shells = np.zeros(extent.size, dtype=int)
+
+    shell_min = -extra_shells
+    shell_max = extent + extra_shells
+
+    # cell coordinates
+    ranges = [slice(lo, hi) for lo, hi in zip(shell_min, shell_max)]
+    # site coordinate within unit cell
+    ranges += [slice(0, len(site_offsets))]
+
+    basis_coords = np.mgrid[ranges].reshape(len(extent) + 1, -1).T
+    positions = basis_coords[:, :-1] @ basis_vectors
+    positions = positions.reshape(-1, len(site_offsets), len(extent)) + site_offsets
+    positions = positions.reshape(-1, len(extent))
+
+    return basis_coords, positions
+
+
+def site_to_idx(basis_coords, extent, site_offsets):
+    """Converts unit cell + sublattice coordinates into lattice site indices.
+
+    basis_coords accepted as an array including sublattice coordinate in last column
+    or as a tuple of unit cell coordinates and a shared sublattice index
+    """
+    if isinstance(basis_coords, Tuple):
+        basis_coords, sl = basis_coords
+    else:
+        basis_coords, sl = basis_coords[:, :-1], basis_coords[:, -1]
+
+    # Accepts extended shells
+    basis_coords = basis_coords % extent
+
+    # Index difference between sites one lattice site apart in each direction
+    # len(site_offsets) for the last axis, as all sites in one cell are listed
+    # factor of extent[-1] to the penultimate axis, etc.
+    radix = np.cumprod([len(site_offsets), *extent[:0:-1]])[::-1]
+
+    return basis_coords @ radix + sl
+
+
+# Near-neighbour search logic
+
+
+def create_padded_sites(basis_vectors, extent, site_offsets, pbc, order):
+    """Generates all lattice sites in an extended shell"""
+    extra_shells = np.where(pbc, order, 0)
+    basis_coords, positions = create_site_positions(
+        basis_vectors, extent, site_offsets, extra_shells
+    )
+    equivalent_ids = site_to_idx(basis_coords, extent, site_offsets)
+
+    return positions, equivalent_ids
+
+
+def get_naive_edges(positions, cutoff, order):
+    """
+    Given an array of spatial `positions`, returns a list `es`, so that
+    `es[k]` contains all pairs of (k + 1)-nearest neighbors up to `order`.
+    Only edges up to distance `cutoff` are considered.
+    """
+    kdtree = cKDTree(positions)
+    dist_matrix = kdtree.sparse_distance_matrix(kdtree, cutoff)
+    row, col, dst = find(triu(dist_matrix))
+    dst = comparable(dst)
+    _, ii = np.unique(dst, return_inverse=True)
+
+    return [sorted(list(zip(row[ii == k], col[ii == k]))) for k in range(order)]
+
+
+def get_nn_edges(
+    basis_vectors,
+    extent,
+    site_offsets,
+    pbc,
+    distance_atol,
+    order,
+):
```

Comment:
can you just write down a brief docstring for this one, like the above functions have? 
And maybe point to `get_custom_edges` as an alternative function doing a similar thing